File: src/acc/acc-vscode/src/downloader.ts
import * as vscode from 'vscode';
import * as https from 'https';
import * as http from 'http';
import * as fs from 'fs';
import * as path from 'path';
import * as tar from 'tar';
import * as child_process from 'child_process';

const ACC_VERSION = '0.3.1'; // Update this with releases
const GITHUB_RELEASES_URL = `https://github.com/KeryxLabs/KeryxInstrumenta/releases/download`;
const ACC_RELEASE_TAG = `acc-engine/v${ACC_VERSION}`;

interface PlatformInfo {
    platform: string;
    arch: string;
    assetName: string;
    binaryName: string;
}

export class AccServerDownloader {
    private context: vscode.ExtensionContext;
    private outputChannel: vscode.OutputChannel;

    constructor(context: vscode.ExtensionContext, outputChannel: vscode.OutputChannel) {
        this.context = context;
        this.outputChannel = outputChannel;
    }

    private getPlatformInfo(): PlatformInfo | null {
        const platform = process.platform;
        const arch = process.arch;

        let assetName: string;
        let binaryName: string;

        if (platform === 'darwin') {
            if (arch === 'arm64') {
                assetName = `acc-${ACC_VERSION}-macos-arm64.tar.gz`;
            } else {
                assetName = `acc-${ACC_VERSION}-macos-x64.tar.gz`;
            }
            binaryName = 'acc';
        } else if (platform === 'linux') {
            if (arch === 'arm64') {
                assetName = `acc-${ACC_VERSION}-linux-arm64.tar.gz`;
            } else {
                assetName = `acc-${ACC_VERSION}-linux-x64.tar.gz`;
            }
            binaryName = 'acc';
        } else if (platform === 'win32') {
            assetName = `acc-${ACC_VERSION}-win-x64.tar.gz`;
            binaryName = 'acc.exe';
        } else {
            return null;
        }

        return { platform, arch, assetName, binaryName };
    }

    public getServerPath(): string | null {
        const config = vscode.workspace.getConfiguration('acc');
        const customPath = config.get<string>('serverPath');
        
        if (customPath) {
            return customPath;
        }

        const platformInfo = this.getPlatformInfo();
        if (!platformInfo) return null;

        const serverDir = path.join(this.context.globalStorageUri.fsPath, 'server');
        const binaryPath = path.join(serverDir, platformInfo.binaryName);

        return fs.existsSync(binaryPath) ? binaryPath : null;
    }

    public getLizardPath(): string | null {
        const config = vscode.workspace.getConfiguration('acc');
        const customPath = config.get<string>('lizardPath');

        if (customPath) {
            return customPath;
        }

        // If lizard is available globally (in PATH), return its name so callers can invoke it.
        if (this.isCommandAvailable('lizard')) {
            return 'lizard';
        }
        return null;
    }

    public async ensureServerInstalled(): Promise<string | null> {
        const existingPath = this.getServerPath();
        if (existingPath) {
            this.outputChannel.appendLine(`ACC server found at: ${existingPath}`);
            return existingPath;
        }

        this.outputChannel.appendLine('ACC server not found, downloading...');

        const result = await vscode.window.showInformationMessage(
            'ACC server binary not found. Download now?',
            'Download',
            'Cancel'
        );

        if (result !== 'Download') {
            return null;
        }

        return await this.downloadServer();
    }

    public async ensureLizardInstalled(): Promise<string | null> {
        const existingPath = this.getLizardPath();
        if (existingPath) {
            this.outputChannel.appendLine(`Lizard found at: ${existingPath}`);
            return existingPath;
        }

        this.outputChannel.appendLine('Lizard not found.');
        const result = await vscode.window.showInformationMessage(
            'Lizard not found. Install via pip?',
            'Install via pip',
            'Cancel'
        );

        if (result === 'Install via pip') {
            const ok = await this.installLizardViaPip();
            if (ok && this.isCommandAvailable('lizard')) {
                return 'lizard';
            }
            vscode.window.showErrorMessage('Failed to install lizard via pip. Install manually and retry.');
            return null;
        }

        return null;
    }

    private installLizardViaPip(): Promise<boolean> {
        return new Promise((resolve) => {
            const pyCandidates = ['python3', 'python'];
            const python = pyCandidates.find(p => this.isCommandAvailable(p));
            if (!python) {
                vscode.window.showErrorMessage('Python not found in PATH. Install Python to use pip.');
                resolve(false);
                return;
            }

            vscode.window.withProgress({
                location: vscode.ProgressLocation.Notification,
                title: 'Installing lizard via pip...',
                cancellable: false
            }, (progress) => {
                return new Promise<void>((innerResolve) => {
                    const args = ['-m', 'pip', 'install', '--upgrade', '--user', 'lizard'];
                    const proc = child_process.spawn(python, args);

                    proc.stdout.on('data', (d) => this.outputChannel.appendLine(d.toString()));
                    proc.stderr.on('data', (d) => this.outputChannel.appendLine(d.toString()));

                    proc.on('close', (code) => {
                        innerResolve();
                        resolve(code === 0);
                    });
                });
            });
        });
    }

    public async downloadServer(): Promise<string | null> {
        const platformInfo = this.getPlatformInfo();
        if (!platformInfo) {
            vscode.window.showErrorMessage(`ACC: Unsupported platform ${process.platform}-${process.arch}`);
            return null;
        }

        const downloadUrl = `${GITHUB_RELEASES_URL}/${ACC_RELEASE_TAG}/${platformInfo.assetName}`;
        const serverDir = path.join(this.context.globalStorageUri.fsPath, 'server');
        
        // Ensure directory exists
        if (!fs.existsSync(serverDir)) {
            fs.mkdirSync(serverDir, { recursive: true });
        }

        const archivePath = path.join(serverDir, platformInfo.assetName);
        const binaryPath = path.join(serverDir, platformInfo.binaryName);

        try {
            await vscode.window.withProgress({
                location: vscode.ProgressLocation.Notification,
                title: "Downloading ACC server...",
                cancellable: false
            }, async (progress) => {
                // Download
                progress.report({ message: 'Downloading binary...' });
                await this.downloadFile(downloadUrl, archivePath, progress);

                // Extract
                progress.report({ message: 'Extracting...' });
                if (platformInfo.assetName.endsWith('.tar.gz')) {
                    await tar.extract({
                        file: archivePath,
                        cwd: serverDir
                    });
                } else {
                    // Handle zip for Windows (you'll need a zip library like 'adm-zip')
                    const AdmZip = require('adm-zip');
                    const zip = new AdmZip(archivePath);
                    zip.extractAllTo(serverDir, true);
                }

                // Make executable on Unix
                if (process.platform !== 'win32') {
                    fs.chmodSync(binaryPath, 0o755);
                }

                // Cleanup archive
                fs.unlinkSync(archivePath);
            });

            this.outputChannel.appendLine(`ACC server installed at: ${binaryPath}`);
            vscode.window.showInformationMessage('ACC server downloaded successfully!');
            
            return binaryPath;

        } catch (err: any) {
            vscode.window.showErrorMessage(`Failed to download ACC server: ${err.message}`);
            this.outputChannel.appendLine(`Download error: ${err}`);
            return null;
        }
    }

    

    private downloadFile(url: string, dest: string, progress: vscode.Progress<{message?: string, increment?: number}>): Promise<void> {
        return new Promise((resolve, reject) => {
            const file = fs.createWriteStream(dest);
            const protocol = url.startsWith('https') ? https : http;

            protocol.get(url, (response) => {
                if (response.statusCode === 302 || response.statusCode === 301) {
                    // Handle redirect
                    file.close();
                    fs.unlinkSync(dest);
                    return this.downloadFile(response.headers.location!, dest, progress)
                        .then(resolve)
                        .catch(reject);
                }

                if (response.statusCode !== 200) {
                    reject(new Error(`Failed to download: ${response.statusCode}`));
                    return;
                }

                const contentType = (response.headers['content-type'] || '').toLowerCase();
                if (contentType.includes('text/html')) {
                    reject(new Error('Download did not return a binary asset (received HTML). Check release tag and asset name.'));
                    return;
                }

                const totalBytes = parseInt(response.headers['content-length'] || '0', 10);
                let downloadedBytes = 0;

                response.on('data', (chunk) => {
                    downloadedBytes += chunk.length;
                    if (totalBytes > 0) {
                        const percent = (downloadedBytes / totalBytes) * 100;
                        progress.report({ 
                            message: `${(downloadedBytes / 1024 / 1024).toFixed(1)} MB / ${(totalBytes / 1024 / 1024).toFixed(1)} MB`,
                            increment: percent 
                        });
                    }
                });

                response.pipe(file);

                file.on('finish', () => {
                    file.close();
                    resolve();
                });

            }).on('error', (err) => {
                fs.unlink(dest, () => {});
                reject(err);
            });

            file.on('error', (err) => {
                fs.unlink(dest, () => {});
                reject(err);
            });
        });
    }

    private isCommandAvailable(cmd: string): boolean {
        try {
            const which = process.platform === 'win32' ? 'where' : 'which';
            const res = child_process.spawnSync(which, [cmd], { encoding: 'utf8' });
            return res.status === 0 && !!res.stdout && res.stdout.trim().length > 0;
        } catch {
            return false;
        }
    }

    
}

File: src/acc/acc-nvim/lua/acc/downloader.lua
-- lua/acc/downloader.lua
-- Handles downloading the ACC engine binary and ensuring lizard is available.

local M = {}

local ACC_VERSION = "0.3.2"
local GITHUB_RELEASES_URL = "https://github.com/KeryxLabs/KeryxInstrumenta/releases/download"
local ACC_RELEASE_TAG = "acc-engine/v" .. ACC_VERSION

-- Returns the data directory where the binary is stored
local function server_dir()
  return vim.fn.stdpath("data") .. "/acc/server"
end

-- Detect platform asset name and binary name
local function platform_info()
  local sysname = vim.loop.os_uname().sysname:lower()
  local machine = vim.loop.os_uname().machine:lower()

  local asset, binary

  if sysname:find("darwin") then
    binary = "acc"
    asset = machine:find("arm") and
      ("acc-" .. ACC_VERSION .. "-macos-arm64.tar.gz") or
      ("acc-" .. ACC_VERSION .. "-macos-x64.tar.gz")
  elseif sysname:find("linux") then
    binary = "acc"
    asset = machine:find("arm") and
      ("acc-" .. ACC_VERSION .. "-linux-arm64.tar.gz") or
      ("acc-" .. ACC_VERSION .. "-linux-x64.tar.gz")
  elseif sysname:find("windows") then
    binary = "acc.exe"
    asset = "acc-" .. ACC_VERSION .. "-win-x64.tar.gz"
  else
    return nil
  end

  return { asset = asset, binary = binary }
end

-- Returns the managed binary path if it exists, otherwise nil
function M.get_server_path(cfg)
  if cfg.server_path and vim.fn.executable(cfg.server_path) == 1 then
    return cfg.server_path
  end

  local info = platform_info()
  if not info then return nil end

  local bin = server_dir() .. "/" .. info.binary
  return vim.fn.filereadable(bin) == 1 and bin or nil
end

-- Returns "lizard" if it's on PATH, otherwise nil
function M.get_lizard_path(cfg)
  if cfg.lizard_path then return cfg.lizard_path end
  return vim.fn.executable("lizard") == 1 and "lizard" or nil
end

-- Download + extract the ACC binary, call cb(path_or_nil) when done
function M.download_server(cb)
  local info = platform_info()
  if not info then
    vim.notify("ACC: Unsupported platform", vim.log.levels.ERROR)
    cb(nil)
    return
  end

  local dir      = server_dir()
  local archive  = dir .. "/" .. info.asset
  local bin_path = dir .. "/" .. info.binary
  local url      = GITHUB_RELEASES_URL .. "/" .. ACC_RELEASE_TAG .. "/" .. info.asset

  vim.fn.mkdir(dir, "p")

  vim.notify("ACC: Downloading engine v" .. ACC_VERSION .. "...", vim.log.levels.INFO)

  -- Use curl (available on all platforms Neovim targets)
  local curl_cmd = { "curl", "-fsSL", "--output", archive, "-L", url }

  vim.fn.jobstart(curl_cmd, {
    on_exit = function(_, code)
      if code ~= 0 then
        vim.notify("ACC: Download failed (curl exit " .. code .. ")", vim.log.levels.ERROR)
        cb(nil)
        return
      end

      vim.notify("ACC: Extracting...", vim.log.levels.INFO)

      local tar_cmd = { "tar", "-xzf", archive, "-C", dir }
      vim.fn.jobstart(tar_cmd, {
        on_exit = function(_, tcode)
          -- Clean up archive regardless
          vim.fn.delete(archive)

          if tcode ~= 0 then
            vim.notify("ACC: Extraction failed", vim.log.levels.ERROR)
            cb(nil)
            return
          end

          -- Make executable on Unix
          if not vim.loop.os_uname().sysname:lower():find("windows") then
            vim.fn.jobstart({ "chmod", "+x", bin_path }, {
              on_exit = function()
                vim.notify("ACC: Engine installed at " .. bin_path, vim.log.levels.INFO)
                cb(bin_path)
              end
            })
          else
            vim.notify("ACC: Engine installed at " .. bin_path, vim.log.levels.INFO)
            cb(bin_path)
          end
        end,
      })
    end,
  })
end

-- Ensure the server binary exists; prompt to download if not
function M.ensure_server_installed(cfg, cb)
  local path = M.get_server_path(cfg)
  if path then
    cb(path)
    return
  end

  vim.ui.select({ "Download", "Cancel" }, {
    prompt = "ACC engine not found. Download v" .. ACC_VERSION .. "?",
  }, function(choice)
    if choice ~= "Download" then
      cb(nil)
      return
    end
    M.download_server(cb)
  end)
end

-- Ensure lizard is available; prompt to pip-install if not
function M.ensure_lizard_installed(cfg, cb)
  local path = M.get_lizard_path(cfg)
  if path then
    cb(path)
    return
  end

  vim.ui.select({ "Install via pip", "Skip" }, {
    prompt = "lizard not found. Install via pip for full functionality?",
  }, function(choice)
    if choice ~= "Install via pip" then
      vim.notify("ACC: lizard not available — some features may be limited", vim.log.levels.WARN)
      cb(nil)
      return
    end

    -- Find python
    local python = vim.fn.executable("python3") == 1 and "python3"
      or vim.fn.executable("python") == 1 and "python"
      or nil

    if not python then
      vim.notify("ACC: Python not found in PATH — install Python to use pip", vim.log.levels.ERROR)
      cb(nil)
      return
    end

    vim.notify("ACC: Installing lizard via pip...", vim.log.levels.INFO)

    vim.fn.jobstart(
      { python, "-m", "pip", "install", "--upgrade", "--user", "lizard" },
      {
        on_exit = function(_, code)
          if code == 0 and vim.fn.executable("lizard") == 1 then
            vim.notify("ACC: lizard installed", vim.log.levels.INFO)
            cb("lizard")
          else
            vim.notify("ACC: lizard install failed — install manually with `pip install lizard`", vim.log.levels.ERROR)
            cb(nil)
          end
        end,
      }
    )
  end)
end

return M
